<a href="https://colab.research.google.com/github/mphill82/Coursera_Capstone/blob/main/Clustering_Toronto_Neighborhoods.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Clustering Toronto Neighborhoods by Venues

In [141]:
import numpy as np
import pandas as pd
import json
import requests
from bs4 import BeautifulSoup

### Part 1 - Acquiring Neighborhood information from Wikipedia

First we scrape data from wikipedia on neighborhoods and boroughs of each postal code in Toronto.

In [199]:
page=requests.get('https://en.wikipedia.org/wiki/List_of_postal_codes_of_Canada:_M').text
soup=BeautifulSoup(page,'html.parser')

In [200]:
table = soup.find('table')

We extract the postal code, borough, and list of associated neighborhoods from each cell of the table on wikipedia and store it to dataframe df.

In [201]:
df=[]

In [202]:
table_contents=[]
table=soup.find('table')
for row in table.findAll('td'):
    cell = {}
    if row.span.text=='Not assigned':
        pass
    else:
        cell['PostalCode'] = row.p.text[:3]   #first three characters in a row
        cell['Borough'] = (row.span.text).split('(')[0]   #everything in the span section of a row before the first
        cell['Neighborhood'] = (((((row.span.text).split('(')[1]).strip(')')).replace(' /',',')).replace(')',' ')).strip(' ')
        table_contents.append(cell)   #starting after the first parentheses, remove the rest of the parentheses and replace forward slashes with commas

df=pd.DataFrame(table_contents)
df['Borough']=df['Borough'].replace({'Downtown TorontoStn A PO Boxes25 The Esplanade':'Downtown Toronto Stn A',
                                             'East TorontoBusiness reply mail Processing Centre969 Eastern':'East Toronto Business',
                                             'EtobicokeNorthwest':'Etobicoke Northwest','East YorkEast Toronto':'East York/East Toronto',
                                             'MississaugaCanada Post Gateway Processing Centre':'Mississauga'})

Finally I sort the dataframe by postalcode and see that there are 103 postal codes.



In [203]:
df=df.sort_values('PostalCode')
df = df.reset_index(drop=True)
df.head()

PostalCode      Borough                            Neighborhood
0        M1B  Scarborough                          Malvern, Rouge
1        M1C  Scarborough  Rouge Hill, Port Union, Highland Creek
2        M1E  Scarborough       Guildwood, Morningside, West Hill
3        M1G  Scarborough                                  Woburn
4        M1H  Scarborough                               Cedarbrae

In [204]:
df.shape

(103, 3)

### Part 2 - Acquiring Latitude and Longitude data

Geocoder was not working so I used the csv file provided by course instructor to access latitude and longitude data corresponding to toronto postal codes.

In [205]:
url='https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/IBMDeveloperSkillsNetwork-DS0701EN-SkillsNetwork/labs_v1/Geospatial_Coordinates.csv'
#Read the file and convert it to a pandas dataframe
df_ll= pd.read_csv(url)
#Combine this dataframe with the neighborhood dataframe from before
df=pd.concat([df,df_ll[['Latitude','Longitude']]],axis=1)
df.head()

PostalCode      Borough                            Neighborhood   Latitude  \
0        M1B  Scarborough                          Malvern, Rouge  43.806686   
1        M1C  Scarborough  Rouge Hill, Port Union, Highland Creek  43.784535   
2        M1E  Scarborough       Guildwood, Morningside, West Hill  43.763573   
3        M1G  Scarborough                                  Woburn  43.770992   
4        M1H  Scarborough                               Cedarbrae  43.773136   

   Longitude  
0 -79.194353  
1 -79.160497  
2 -79.188711  
3 -79.216917  
4 -79.239476

In [206]:
df.shape

(103, 5)

## Part 3 - Clustering with Foursquare Venue data

In [143]:
import folium
import geopy

Here we find the location of the center of Toronto so we can create a map of the city.  We are going to display all the postal code locations on the map.

In [144]:
address = 'Toronto, ON'

geolocator = Nominatim(user_agent="ny_explorer")
location = geolocator.geocode(address)
latitude = location.latitude
longitude = location.longitude
print('The geograpical coordinate of Toronto, ON are {}, {}.'.format(latitude, longitude))

The geograpical coordinate of Toronto, ON are 43.6534817, -79.3839347.


### All the postal code locations of Toronto

In [207]:
# create map of Toronto using latitude and longitude values
f = folium.Figure(width=650, height=450)
map_toronto = folium.Map(location=[latitude, longitude], zoom_start=10).add_to(f)

# add markers to map
for lat, lng, borough, pc in zip(df['Latitude'], df['Longitude'], df['Borough'], df['PostalCode']):
    label = '{}, {}'.format(pc, borough)
    label = folium.Popup(label, parse_html=True)
    folium.CircleMarker(
        [lat, lng],
        radius=5,
        popup=label,
        color='blue',
        fill=True,
        fill_color='#3186cc',
        fill_opacity=0.7,
        parse_html=False).add_to(map_toronto)  
    
map_toronto

#### We will restrict our analysis only to boroughs that contain the name Toronto.  Boroughs that don't meet this condition don't have enough venues to be relevant to our clustering.

In [209]:
df_toronto = df[df['Borough'].apply(lambda x: True if 'Toronto' in x else False)]
df_toronto.head()

PostalCode                 Borough                    Neighborhood  \
37        M4E            East Toronto                     The Beaches   
40        M4J  East York/East Toronto              The Danforth  East   
41        M4K            East Toronto    The Danforth West, Riverdale   
42        M4L            East Toronto  India Bazaar, The Beaches West   
43        M4M            East Toronto                 Studio District   

     Latitude  Longitude  
37  43.676357 -79.293031  
40  43.685347 -79.338106  
41  43.679557 -79.352188  
42  43.668999 -79.315572  
43  43.659526 -79.340923

In [211]:
# create map of Toronto using latitude and longitude values
f = folium.Figure(width=650, height=450)
map_toronto = folium.Map(location=[latitude, longitude], zoom_start=11).add_to(f)

# add markers to map
for lat, lng, borough, pc in zip(df_toronto['Latitude'], df_toronto['Longitude'], df_toronto['Borough'], df_toronto['PostalCode']):
    label = '{}, {}'.format(pc, borough)
    label = folium.Popup(label, parse_html=True)
    folium.CircleMarker(
        [lat, lng],
        radius=5,
        popup=label,
        color='blue',
        fill=True,
        fill_color='#3186cc',
        fill_opacity=0.7,
        parse_html=False).add_to(map_toronto)  
    
map_toronto

### Define foursquare credentials

In [158]:
CLIENT_ID = '2M2FVW3D5I4QK0ZPX0XWRRECY3XUU5JQK3O5V0GV2LX0MR3Q' # your Foursquare ID
CLIENT_SECRET = '4WE2VPK1JZQN0R4MH3FRMWJA25KNB5VVXM1MVA4BCS2SZ1RI' # your Foursquare Secret
VERSION = '20180605' # Foursquare API version
LIMIT = 100 # A default Foursquare API limit value

print('Your credentails:')
print('CLIENT_ID: ' + CLIENT_ID)
print('CLIENT_SECRET:' + CLIENT_SECRET)

Your credentails:
CLIENT_ID: 2M2FVW3D5I4QK0ZPX0XWRRECY3XUU5JQK3O5V0GV2LX0MR3Q
CLIENT_SECRET:4WE2VPK1JZQN0R4MH3FRMWJA25KNB5VVXM1MVA4BCS2SZ1RI


### Let's create a function to find all the nearby venues of any area of Toronto.

In [178]:
def getNearbyVenues(postalcodes, latitudes, longitudes, radius=500):
    
    venues_list=[]
    for pc, lat, lng in zip(postalcodes, latitudes, longitudes):
            
        # create the API request URL
        url = 'https://api.foursquare.com/v2/venues/explore?&client_id={}&client_secret={}&v={}&ll={},{}&radius={}&limit={}'.format(
            CLIENT_ID, 
            CLIENT_SECRET, 
            VERSION, 
            lat, 
            lng, 
            radius, 
            LIMIT)
            
        # make the GET request
        results = requests.get(url).json()["response"]['groups'][0]['items']
        
        # return only relevant information for each nearby venue
        venues_list.append([(
            pc, 
            lat, 
            lng, 
            v['venue']['name'], 
            v['venue']['location']['lat'], 
            v['venue']['location']['lng'],  
            v['venue']['categories'][0]['name']) for v in results])

    nearby_venues = pd.DataFrame([item for venue_list in venues_list for item in venue_list])
    nearby_venues.columns = ['Postal Code', 
                  'PC Latitude', 
                  'PC Longitude', 
                  'Venue', 
                  'Venue Latitude', 
                  'Venue Longitude', 
                  'Venue Category']
    
    return(nearby_venues)

In [212]:
toronto_venues=getNearbyVenues(df_toronto['PostalCode'],df_toronto['Latitude'],df_toronto['Longitude'])
toronto_venues.head()

Postal Code  PC Latitude  PC Longitude                               Venue  \
0         M4E    43.676357    -79.293031                   Glen Manor Ravine   
1         M4E    43.676357    -79.293031  The Big Carrot Natural Food Market   
2         M4E    43.676357    -79.293031                 Grover Pub and Grub   
3         M4E    43.676357    -79.293031                       Upper Beaches   
4         M4J    43.685347    -79.338106                        Aldwych Park   

   Venue Latitude  Venue Longitude     Venue Category  
0       43.676821       -79.293942              Trail  
1       43.678879       -79.297734  Health Food Store  
2       43.679181       -79.297215                Pub  
3       43.680563       -79.292869       Neighborhood  
4       43.684901       -79.341091               Park

In [213]:
toronto_venues.shape

(1580, 7)

We've collected 1580 venues and identified what area they belong to as well as what type venue they are.

We can look at how many vanues belong to each area...

In [214]:
toronto_venues.groupby('Postal Code')['Venue'].count().to_frame().T

Postal Code  M4E  M4J  M4K  M4L  M4M  M4N  M4P  M4R  M4S  M4T  M4V  M4W  M4X  \
Venue          4    4   42   20   38    4    9   19   36    1   14    4   45   

Postal Code  M4Y  M5A  M5B  M5C  M5E  M5G  M5H  M5J  M5K  M5L  M5N  M5P  M5R  \
Venue         77   45  100   81   60   62   94  100  100  100    3    4   20   

Postal Code  M5S  M5T  M5V  M5W  M5X  M6G  M6H  M6J  M6K  M6P  M6R  M6S  M7Y  
Venue         32   62   14   98  100   16   16   42   24   25   15   33   17

We can also look at how many unique venue categories we have...

In [215]:
print('There are {} uniques categories.'.format(len(toronto_venues['Venue Category'].unique())))

There are 230 uniques categories.


## Pre-processing

We're going to get our data ready for clustering postal code areas by nearby venue type (category).  We'll do this by adding a feature column to our dataframe for each unique venue category.

In [216]:
# one hot encoding
toronto_onehot = pd.get_dummies(toronto_venues[['Venue Category']], prefix="", prefix_sep="")
toronto_df=pd.concat([toronto_venues[['Postal Code']],toronto_onehot],axis=1)
toronto_df.head()

Postal Code  Adult Boutique  Airport  Airport Food Court  Airport Gate  \
0         M4E               0        0                   0             0   
1         M4E               0        0                   0             0   
2         M4E               0        0                   0             0   
3         M4E               0        0                   0             0   
4         M4J               0        0                   0             0   

   Airport Lounge  Airport Service  Airport Terminal  American Restaurant  \
0               0                0                 0                    0   
1               0                0                 0                    0   
2               0                0                 0                    0   
3               0                0                 0                    0   
4               0                0                 0                    0   

   Antique Shop  Aquarium  Art Gallery  Art Museum  Arts & Crafts Store  \
0             0         0            0           0                    0   
1             0         0            0           0                    0   
2             0         0            0           0                    0   
3             0         0            0           0                    0   
4             0         0            0           0                    0   

   Asian Restaurant  Athletics & Sports  BBQ Joint  Baby Store  Bagel Shop  \
0                 0                   0          0           0           0   
1                 0                   0          0           0           0   
2                 0                   0          0           0           0   
3                 0                   0          0           0           0   
4                 0                   0          0           0           0   

   Bakery  Bank  Bar  Baseball Stadium  Basketball Stadium  Beach  \
0       0     0    0                 0                   0      0   
1       0     0    0                 0                   0      0   
2       0     0    0                 0                   0      0   
3       0     0    0                 0                   0      0   
4       0     0    0                 0                   0      0   

   Bed & Breakfast  Beer Bar  Beer Store  Belgian Restaurant  Bistro  \
0                0         0           0                   0       0   
1                0         0           0                   0       0   
2                0         0           0                   0       0   
3                0         0           0                   0       0   
4                0         0           0                   0       0   

   Boat or Ferry  Bookstore  Boutique  Brazilian Restaurant  Breakfast Spot  \
0              0          0         0                     0               0   
1              0          0         0                     0               0   
2              0          0         0                     0               0   
3              0          0         0                     0               0   
4              0          0         0                     0               0   

   Brewery  Bubble Tea Shop  Building  Burger Joint  Burrito Place  Bus Line  \
0        0                0         0             0              0         0   
1        0                0         0             0              0         0   
2        0                0         0             0              0         0   
3        0                0         0             0              0         0   
4        0                0         0             0              0         0   

   Business Service  Butcher  Café  Cajun / Creole Restaurant  Candy Store  \
0                 0        0     0                          0            0   
1                 0        0     0                          0            0   
2                 0        0     0                          0            0   
3                 0        0     0            

Now we can group this dataframe by postal code area and take the mean of each feature column.  This will give us a frequency of occurance for each venue category.

In [217]:
toronto_grouped=toronto_df.groupby('Postal Code').mean().reset_index()
toronto_grouped.head()

Postal Code  Adult Boutique  Airport  Airport Food Court  Airport Gate  \
0         M4E             0.0      0.0                 0.0           0.0   
1         M4J             0.0      0.0                 0.0           0.0   
2         M4K             0.0      0.0                 0.0           0.0   
3         M4L             0.0      0.0                 0.0           0.0   
4         M4M             0.0      0.0                 0.0           0.0   

   Airport Lounge  Airport Service  Airport Terminal  American Restaurant  \
0             0.0              0.0               0.0             0.000000   
1             0.0              0.0               0.0             0.000000   
2             0.0              0.0               0.0             0.023810   
3             0.0              0.0               0.0             0.000000   
4             0.0              0.0               0.0             0.052632   

   Antique Shop  Aquarium  Art Gallery  Art Museum  Arts & Crafts Store  \
0           0.0       0.0          0.0         0.0                  0.0   
1           0.0       0.0          0.0         0.0                  0.0   
2           0.0       0.0          0.0         0.0                  0.0   
3           0.0       0.0          0.0         0.0                  0.0   
4           0.0       0.0          0.0         0.0                  0.0   

   Asian Restaurant  Athletics & Sports  BBQ Joint  Baby Store  Bagel Shop  \
0               0.0                 0.0        0.0         0.0         0.0   
1               0.0                 0.0        0.0         0.0         0.0   
2               0.0                 0.0        0.0         0.0         0.0   
3               0.0                 0.0        0.0         0.0         0.0   
4               0.0                 0.0        0.0         0.0         0.0   

     Bakery      Bank       Bar  Baseball Stadium  Basketball Stadium  Beach  \
0  0.000000  0.000000  0.000000               0.0                 0.0    0.0   
1  0.000000  0.000000  0.000000               0.0                 0.0    0.0   
2  0.023810  0.000000  0.000000               0.0                 0.0    0.0   
3  0.000000  0.000000  0.000000               0.0                 0.0    0.0   
4  0.052632  0.026316  0.026316               0.0                 0.0    0.0   

   Bed & Breakfast  Beer Bar  Beer Store  Belgian Restaurant  Bistro  \
0              0.0       0.0         0.0                 0.0     0.0   
1              0.0       0.0         0.0                 0.0     0.0   
2              0.0       0.0         0.0                 0.0     0.0   
3              0.0       0.0         0.0                 0.0     0.0   
4              0.0       0.0         0.0                 0.0     0.0   

   Boat or Ferry  Bookstore  Boutique  Brazilian Restaurant  Breakfast Spot  \
0            0.0   0.000000       0.0                   0.0             0.0   
1            0.0   0.000000       0.0                   0.0             0.0   
2            0.0   0.023810       0.0                   0.0             0.0   
3            0.0   0.000000       0.0                   0.0             0.0   
4            0.0   0.026316       0.0                   0.0             0.0   

    Brewery  Bubble Tea Shop  Building  Burger Joint  Burrito Place  Bus Line  \
0  0.000000          0.00000       0.0           0.0            0.0       0.0   
1  0.000000          0.00000       0.0           0.0            0.0       0.0   
2  0.023810          0.02381       0.0           0.0            0.0       0.0   
3  0.050000          0.00000       0.0           0.0            0.0       0.0   
4  0.052632          0.00000       0.0           0.0            0.0       0.0   

   Business Service  Butcher      Café  Cajun / Creole Restaurant  \
0               0.0      0.0  0.000000                        0.0   
1               0.0      0.0  0.000000                        0.0   
2               0.0      0.0  0.023810                        0.0   
3         

### Let's print each postal code area along with the top 5 most common venue types.

In [218]:
num_top_venues = 5

for hood in toronto_grouped['Postal Code']:
    print("----"+hood+"----")
    temp = toronto_grouped[toronto_grouped['Postal Code'] == hood].T.reset_index()   #transpose will list the venue categories as columns
    temp.columns = ['venue','freq']
    temp = temp.iloc[1:]    #remove the first row because its "Neighborhood" , "Battery Park City"
    temp['freq'] = temp['freq'].astype(float)
    temp = temp.round({'freq': 2})
    print(temp.sort_values('freq', ascending=False).reset_index(drop=True).head(num_top_venues))
    print('\n')

----M4E----
               venue  freq
0                Pub  0.25
1  Health Food Store  0.25
2       Neighborhood  0.25
3              Trail  0.25
4     Adult Boutique  0.00


----M4J----
               venue  freq
0               Park  0.50
1  Convenience Store  0.25
2       Intersection  0.25
3     Adult Boutique  0.00
4             Museum  0.00


----M4K----
                    venue  freq
0        Greek Restaurant  0.19
1             Coffee Shop  0.10
2      Italian Restaurant  0.07
3  Furniture / Home Store  0.05
4          Ice Cream Shop  0.05


----M4L----
                  venue  freq
0                  Park  0.10
1  Fast Food Restaurant  0.10
2    Italian Restaurant  0.05
3             Pet Store  0.05
4           Coffee Shop  0.05


----M4M----
                 venue  freq
0          Coffee Shop  0.08
1              Brewery  0.05
2                 Café  0.05
3  American Restaurant  0.05
4               Bakery  0.05


----M4N----
              venue  freq
0              Park  0